gateway/ln-gateway/src/lnd.rs

```
@@ -0,0 +1,302 @@
+        })
+        .await;
+
+        Ok(Box::pin(ReceiverStream::new(gwd_rx)))
+    }
+
+    async fn complete_htlc(
+        &self,
+        outcome: CompleteHtlcsRequest,
+    ) -> crate::Result<CompleteHtlcsResponse> {
+        let CompleteHtlcsRequest {
+            action,
+            intercepted_htlc_id,
+        } = outcome;
+
+        let hash = match sha256::Hash::from_slice(&intercepted_htlc_id) {
+            Ok(hash) => hash,
+            Err(e) => {
+                error!("Invalid intercepted_htlc_id: {:?}", e);
+                return Err(GatewayError::LnRpcError(tonic::Status::invalid_argument(
+                    e.to_string(),
+                )));
+            }
+        };
+
+        info!("Completing htlc with reference, {:?}", hash);
+
+        if let Some((outcome, incoming_circuit_key)) = self.outcomes.lock().await.remove(&hash) {
+            let htlca_res = match action {
+                Some(Action::Settle(Settle { preimage })) => ForwardHtlcInterceptResponse {
+                    incoming_circuit_key,
+                    action: 0,
+                    preimage,
+                    failure_message: vec![],
+                    failure_code: 0,
+                },
+                Some(Action::Cancel(Cancel { reason: _ })) => cancel_intercepted_htlc(None),
+                None => {
+                    error!("No action specified for intercepted htlc id: {:?}", hash);
+                    return Err(GatewayError::LnRpcError(tonic::Status::internal(
+                        "No action specified on this intercepted htlc",
+                    )));
+                }
+            };
+
+            // TODO: Consider retrying this if the send fails
+            let _ = outcome.send(htlca_res).await.map_err(|_| {
+                GatewayError::LnRpcError(tonic::Status::internal(
+                    "Failed to send htlc processing outcome to LND node",
+                ))
+            });
```

Don't suppress this error.

---

done